Clews 14263 part2 analogous years

joysaikat opened the pull request.

Commit:
Added files to api-client directory

joysaikat committed:
bringing the analogous year package to github

joysaikat committed:
Added matplotlib in the requirements folder

joysaikat committed:
Modified the requirements file and updated the exception handling in case of a empty dataframe return

joysaikat committed:
Importing combinatorial function from scipy.special instead of scipy.misc

joysaikat committed:
added timezones after parsing datetime strings

joysaikat committed:
edits and reformats

joysaikat committed:
python 2.7 and matplotlib.pyplot had issues https://stackoverflow.com/questions/36327134/matplotlib-error-no-module-named-tkinter . Trying a solution laid out in the link

joysaikat committed:
Resolved issues with infinites showing up in extracted features for very small intervals

joysaikat committed:
Fixed a unit test fail

joysaikat committed:
Added conditions for better timezone awareness handling across different versions of pandas

joysaikat committed:
Formatting

joysaikat committed:
Fixed a wrong import

joysaikat committed:
Corrected the example command in readme.md

tolothy commented on api/client/samples/analogous_years/lib/final_ranks_computation.py:
```diff
@@ -0,0 +1,188 @@
+import pandas as pd
+from scipy.special import comb
+
+from . import \
```
let's try absolute path import

joysaikat replied:
done

tolothy commented on api/client/samples/analogous_years/lib/final_ranks_computation_test.py:
```diff
@@ -0,0 +1,29 @@
+import pandas as pd
+from pandas.util.testing import assert_frame_equal
+
+from .final_ranks_computation import combined_methods_distances
```
absolute import

joysaikat replied:
changed

tolothy commented on api/client/samples/analogous_years/lib/get_transform_data_test.py:
```diff
@@ -0,0 +1,227 @@
+from pandas.util.testing import assert_frame_equal
+import pytest
+
+from . import get_transform_data
```
absolute import

joysaikat replied:
changed

joysaikat committed:
changed relative imports to absolute imports

joysaikat committed:
Removing all logical operation out of main(); In this iteration moved enso and start_date related logics

joysaikat committed:
Removed more logic from main functions related to naming the stored report

joysaikat committed:
moved start date computation out of the run_analogous_years script

joysaikat committed:
changed an import statement

joysaikat committed:
Completely moved logical computations outside of the main() function in run_analogous_years.py

joysaikat committed:
Fixed an import statement for remote branch

joysaikat committed:
Merge branch 'development' into CLEWS-14263-Part2-analogous_years

joysaikat committed:
add analogous_years files

joysaikat committed:
Simplified the combined_items_final_ranks function further

joysaikat committed:
Added code to remove properties from entity other than, item_id, metric_id, region_id, source_id, frequency_id, start_date

joysaikat committed:
Fixed a runtime error faced by the previous code due to iterating and deleting over keys of a dictionary

joysaikat committed:
Main change is to readme file; additionally changed how the import date is resampled

JGrinst commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,158 @@
+
+## Working Example Using Python
+To get started with `Analogous Years`, users have to install `Gro API Client`
+as detailed [here](https://github.com/gro-intelligence/api-client). 
```
We can link to our official docs page instead: https://developers.gro-intelligence.com/installation.html

joysaikat replied:
Thanks, updated!

tolothy reviewed (commented):
could you delete all the __pychache__ dir and all .pyc files?

JGrinst commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,158 @@
+# Analogous Years
+Users can find similarity based ranks between a time period and a list of time periods 
+given by the time period ± 1 year, ± 2 years, ..., ± N years using `Analogous Years` application.
+For rank computation a user must provide:
```
I think the intro should be a bit more accessible.  Perhaps something like:

"The `Analogous Years` application enables users to compare events from a set period of time, to those of the same date range in other years. The application will compute ranks of similarity between the specified period, and the same period from previous or future years. Multiple different inputs can be used in determining the ranks, but to compute these ranks, a user must provide:

joysaikat replied:
Copying your lines, thanks!

joysaikat committed:
deleted unnecessary files

JGrinst commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,158 @@
+If the user wants to know which period of time is closest to the time period 
+between 1<sup>st</sup> January 2019 and 31<sup>st</sup> October 2019 with respect to the
+following Gro-entities -
+1. Rainfall (metric_id=2100031, item_id=2039, source_id=35, frequency_id=1)
```
We actually have multiple different rainfall sources, so it might be a bit misleading to say this represents rainfall. Source would need to be specified in the name, as well.

My inner Nemo voice also has an aversion to setting ID numbers. But it's probably fine here.

joysaikat replied:
Added the names of the sources associated with each entity.

joysaikat committed:
Removed __pycache__ files

JGrinst commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,158 @@
+To get started with `Analogous Years`, users have to install `Gro API Client`
+as detailed [here](https://github.com/gro-intelligence/api-client). 
+
+If the user wants to know which period of time is closest to the time period 
```
which period of time is closest => which period of time is most similar

?

joysaikat replied:
Done, thanks!

JGrinst commented on api/client/samples/analogous_years/readme.md:
````diff
@@ -0,0 +1,158 @@
+analogous_years $ python run_analogous_years.py -m 2100031 2540047 15531082 -i 2039 3457 7382 -r 100000100 -s 35 26 43 -f 1 1 1 --initial_date 2019-01-01 --final_date 2019-10-31
+```
+The ranks will be saved as a csv file in the same directory.
+## Appendix
````
Add an extra line here, before appendix

joysaikat replied:
Done

joysaikat committed:
addressed comments by Justin

tolothy reviewed (approved):
Look good to me now

joysaikat committed:
Updated the name of the main function

joysaikat committed:
minor edits

joysaikat committed:
adding a jupyter notebook to get started with analogous years

joysaikat committed:
minor edit to output composite rank as dataframe

nemozen commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,164 @@
+2. Land Temperature, MODIS (metric_id=2540047, item_id=3457, source_id=26, frequency_id=1)
+3. Soil moisture, SMOS (metric_id=15531082, item_id=7382, source_id=43, frequency_id=1)
+
+in the US Corn Belt (region_id=100000100) with equal weights given (by default) to 
```
try to link these entities to dictionary https://app.gro-intelligence.com/dictionary/regions/100000100

joysaikat replied:
Done

nemozen commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,164 @@
+
+# Rainfall (modeled) - Precipitation Quantity - 
+# US Corn Belt States (NASA TRMM 3B42RT)
+entity_1 = {'metric_id': 2100031, 
```
data_series or selection

joysaikat replied:
Done

nemozen commented on api/client/samples/analogous_years/lib/final_ranks_computation.py:
```diff
@@ -0,0 +1,259 @@
+    return file_name, display_dataframe
+
+
+def save_to_csv(dataframe_file_name, logger, all_ranks=None, report=True, output_dir=''):
```
csv and png output should be two different functions

joysaikat replied:
Done

nemozen commented on api/client/samples/analogous_years/lib/final_ranks_computation.py:
```diff
@@ -0,0 +1,259 @@
+            weights.append(enso_weight)
+        else:
+            weights.append(1)
+    file_name = get_file_name(client, entities, initial_date, final_date)
```
this is nicely separable so keep it outside
and make this function only return the dataframe

joysaikat replied:
Done

nemozen commented on api/client/samples/analogous_years/lib/get_transform_data.py:
```diff
@@ -0,0 +1,140 @@
+                                     'source_id': source_id,
+                                     'frequency_id': frequency_id,
+                                     'start_date': start_date})
+    data = pd.DataFrame(data)
```
use

```
for series in the series you wnat:
   client.add_single_data_series(** series)

client.get_df()
```
this will give you a dataframe that contains the

joysaikat replied:
I have added get_df in the current workflow. 
Note: Assuming series is a dictionary of the form `series =  {'metric_id': metric_id, 'item_id': item_id, 'region_id': region_id, 'frequency_id': frequency_id, 'source_id': source_id, 'start_date': start_date}`, then the code would look like
```
client.add_single_data_series(series)
client.get_df()
```

nemozen commented on api/client/samples/analogous_years/readme.md:
```diff
@@ -0,0 +1,164 @@
+# Analogous Years
```
let's remove redundant info between readme and notebook now that the latter exists
readme can jus tbe a short cmd line example
and then can point to notebook for more complex usage

joysaikat replied:
There is still some overlap. However, I have removed much of it.

nemozen reviewed (approved):
when ready to run the final notebook, you can edit 
api.client.cfg.py and change log level to ERROR

joysaikat committed:
Corrected a naming convention: Replaced entity with data_series where necessary

joysaikat committed:
Corrected a naming convention: Replaced entity with data_series where necessary

joysaikat committed:
Simplified some of the output functions by separating the output dataframe from saving .csv file, .png file

joysaikat committed:
Using seaborn for plotting instead of matplotlib

joysaikat committed:
Formatting in ipynb, readme.md

joysaikat committed:
Adding seaborn to api-client/requirements-test.txt

joysaikat committed:
trying to resolve a test failure in shippable

joysaikat committed:
trying to force matplotlib to use a proper backend

joysaikat committed:
removed unnecessary function from main (logger)

joysaikat committed:
Included get_df() and fixed the corresponding unit test

nemozen commented on api/.gitignore:
```diff
@@ -1,2 +1,3 @@
-*.csv
\ No newline at end of file
+*.csv
+.idea/
```
revert this unrelated change

joysaikat replied:
Done

nemozen commented on requirements-test.txt:
```diff
@@ -1,5 +1,12 @@
+dtw==1.3.3
+matplotlib
 mock
-sklearn
```
why change ? https://stackoverflow.com/questions/38733220/difference-between-scikit-learn-and-sklearn

joysaikat replied:
To install `scikit-learn` using pip we have to run ```pip install scikit-learn```. That is why I changed it in the requirements-test.txt (assuming that package will be installed during testing)

https://scikit-learn.org/0.16/install.html

To import `scikit-learn` the code is ```import sklearn```

itsjohnward replied:
`sklearn`'s an alias in PyPI for "latest version of `scikit-learn`" so version pinning doesn't work with it. Also, the alias does not exist in conda. Neither issue matters here in particular since this file is just for our unit tests, but in general `scikit-learn` is the better all-around name to use

joysaikat replied:
Right, I just tried ```pip install sklearn```, it seems to work. However, I clearly remember doing ```pip install sklearn==<>``` that did not work.

nemozen reviewed (changes requested):
couple of very minor comments.

also please remove the warning about  "Panel class is removed from pandas:" in the noteboook... you can either edit it out manually or rerun the notebook with log level set to ERROR in cfg.py

joysaikat committed:
reverted an unnecessary change + removed a warning manually from the ipynb notebook

joysaikat committed:
Changed some code to address SettingWithCopyWarning (did not work)

joysaikat committed:
Small changes in markdown

nemozen commented:
almost there!  the notebook is still has a little bit of cruft that's not ideal for a show case. 

- can you remove the commented out code at the end?
- the table of contents is not that useful, and the links don't seem to work.. can you just remove it.
- the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.
- does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually

joysaikat committed:
Formatting changes for the jupyter notebook

joysaikat commented:
> almost there! the notebook is still has a little bit of cruft that's not ideal for a show case.
> 
> * can you remove the commented out code at the end?

Kept there as plotly produces cooler visualizations, but plotly plots could not be rendered in github. So, commented out if users wanted to download and use the notebook locally. Removed, in this version.

> * the table of contents is not that useful, and the links don't seem to work.. can you just remove it.

Discussed in slack. The links do not work because of github. They work fine if the notebook is downloaded locally or nbviewer is used. However, it makes the notebook longer. So, removed it.

> * the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.

Removed it.
> * does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually

Discussed in slack. I understand the need for a more concise notebook; hence, removed it. However, manually removing is a set back for reproducibility of the notebook.

joysaikat committed:
reverted back some changes in value assinment, no longer using .loc

nemozen commented:
> > almost there! the notebook is still has a little bit of cruft that's not ideal for a show case.
> > 
> > * can you remove the commented out code at the end?
> 
> Kept there as plotly produces cooler visualizations, but plotly plots could not be rendered in github. So, commented out if users wanted to download and use the notebook locally. Removed, in this version.
> 
> > * the table of contents is not that useful, and the links don't seem to work.. can you just remove it.
> 
> Discussed in slack. The links do not work because of github. They work fine if the notebook is downloaded locally or nbviewer is used. However, it makes the notebook longer. So, removed it.
> 
> > * the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.
> 
> Removed it.
> 
> > * does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually
> 
> Discussed in slack. I understand the need for a more concise notebook; hence, removed it. However, manually removing is a set back for reproducibility of the notebook.

the user should be able to choose different log verbosity levels  when running it.  forcing the log level in code (with logging.disable) is bad practice because it would suppress other third party  warnings that the user may actually encounter (that may or may not occur in your case), even if the user wants to see warnings.   

more importantly, please don't forget to follow-up to address the underlying causes of the warnings. in the meantime, it's not helpful to have them pollute the notebook
